src/frequency.rs

```diff
@@ -78,6 +78,11 @@ impl<T: Eq + Hash> Frequencies<T> {
     pub fn len(&self) -> usize {
         self.data.len()
     }
+
+    /// Return an iterator over the unique values of the data.
+    pub fn unique_values(&self) -> impl Iterator<Item=&T> {
+        self.data.keys()
+    }
```

Comment:
In principle I'm okay with this, but you'll need to define a new concrete iterator type. Unless there's a specific reason to do so, it is bad juju to return an `impl Trait` in a public API.

Reply:
Sure. Would the concrete type [`std::collections::hash_map::Keys` ](https://doc.rust-lang.org/std/collections/hash_map/struct.Keys.html) work, or should I wrap it in another `struct`? 

I'm pretty curious though why you think having `impl Trait` in public API is a bad idea (other than bumping the minimum supported Rust version), because I thought that was the whole point of this feature. Could you maybe point me to discussion or a style guide rationale about this somewhere?

Reply:
> Sure. Would the concrete type `std::collections::hash_map::Keys` work, or should I wrap it in another struct?

I'd rather it be a new type.

> I'm pretty curious though why you think having `impl Trait` in public API is a bad idea (other than bumping the minimum supported Rust version), because I thought that was the whole point of this feature. Could you maybe point me to discussion or a style guide rationale about this somewhere?

To clarify, not every opinion I have about Rust idioms or styles can be backed up by a thorough discussion or style guide.

A key qualifier in what I said is the word public. The whole point of `impl Trait` is completely orthogonal to public APIs. The reason why using it as a return type in public APIs is problematic is because iterator types often implement several traits, such as `Clone` or `Debug`. (`hash_map::Keys` actually implements a couple more than that). But by restricting the return type to just `impl Iterator`, then the only impl you get is the `Iterator` impl. You lose `Debug`, Clone` and whatever else there might be. That's not good. On the contrary, using the standard newtype pattern permits adding additional impls, or even doing it later. Finally, on top of all of that, `impl Trait` is only permitted in very constrained use cases, which makes it tricky to compose. For example, you cannot put it into another type definition at the moment.

`impl Trait` is useful as an internal convenience, and in particular, futures-oriented code is a place where this comes in handy because of the amount of ad hoc internal composition one might do. `impl Trait` permits splitting that composition over function boundaries in a fairly frictionless way, which is valuable. None of this means that `impl Trait` should be sprinkled everywhere. :-)

I don't think the Rust API guidelines have caught up to this yet, but I'd expect something like the above to be written for it if an entry were added today.

Reply:
Thank you for the thorough explanation. Those are indeed some good points, I'll bear them in mind the next time I consider using `impl Trait` in public API :)

I've created the new type. Please let me know if there are any issues with the code style or names.